### PR TITLE
enhancement(observability): Enable runtime allocation tracking by default. 

### PR DESCRIPTION
**Pull request by arshiyasolei**
This PR enables support for allocation tracing by default (via the `allocation-tracing` feature). Users can enable it by passing the `--allocation-tracing` command-line flag, or setting the `ALLOCATION_TRACING` environment variable to `true`.

### CODE REVIEW

**Comment by arshiyasolei** [comment]
Notes: Passing a smaller than expected layout is UB

File: src/internal_telemetry/allocations/allocator/tracing_allocator.rs [hunk 1]
```diff
@@ -48,40 +48,46 @@ impl<A: GlobalAlloc, T: Tracer> GroupedTraceableAllocator<A, T> {
...
             try_with_suspended_allocation_group(
                 #[inline(always)]
                 |group_id| {
                     group_id_ptr.write(group_id.as_raw());
                     self.tracer.trace_allocation(object_size, group_id);
                 },
             );
+            object_ptr
+        } else {
+            self.allocator.alloc(object_layout)
```
Comment (tobz) [inline src/internal_telemetry/allocations/allocator/tracing_allocator.rs:71]: I think it would be better to structure this code as:

```
if !tracking_allocations {
  return self.allocator.alloc(...)
}

<do a tracked allocation>
```

Right now, it's a little hard to follow the logic.

File: src/internal_telemetry/allocations/allocator/tracing_allocator.rs [hunk 2]
```diff
@@ -48,40 +48,46 @@ impl<A: GlobalAlloc, T: Tracer> GroupedTraceableAllocator<A, T> {
...
-        let object_size = object_layout.size();
-        // Group id value of zero implies allocations tracking was disabled
-        // during this allocation. We override this if allocations were in fact enabled.
-        group_id_ptr.write(0);
         if TRACK_ALLOCATIONS.load(Ordering::Relaxed) {
+            let (group_id_ptr, object_ptr, _wrapped_layout) =
+                self.get_wrapped_allocation(object_layout);
+            let object_size = object_layout.size();
+            // Group id value of zero implies allocations tracking was disabled
+            // during this allocation. We override this if allocations were in fact enabled.
```
Comment (tobz) [inline src/internal_telemetry/allocations/allocator/tracing_allocator.rs:56]: This comment, and the one in `dealloc` that got removed around the `if group_id == 0` line, are now out-of-date. If we're here at this line, setting the group ID to zero, we are defacto tracking allocations, even if it's technically only startup.

File: src/app.rs [hunk 3]
```diff
@@ -61,6 +61,34 @@ pub struct Application {
...
+            if allocation_tracing {
+                // Configure our tracking allocator.
+                init_allocation_tracing();
+                TRACK_ALLOCATIONS.store(true, Ordering::Relaxed);
+            } else {
+                // We assume nothing is allocated on the heap at this point.
+                TRACK_ALLOCATIONS.store(false, Ordering::Relaxed);
+            }
+        }
+
```
Comment (tobz) [inline src/app.rs:91]: I think if we want to be sure we're calculating this as early as possible, it should actually be at the _very_ top of `main`, where we were calling `vector::internal_telemetry::allocations::init_allocation_tracing()` before.

File: src/internal_telemetry/allocations/allocator/tracing_allocator.rs [hunk 4]
```diff
@@ -96,6 +102,17 @@ unsafe impl<A: GlobalAlloc, T: Tracer> GlobalAlloc for GroupedTraceableAllocator
...
+    #[inline]
+    unsafe fn realloc(&self, ptr: *mut u8, layout: Layout, new_size: usize) -> *mut u8 {
+        let new_layout = Layout::from_size_align_unchecked(new_size, layout.align());
+        let new_ptr = self.alloc(new_layout);
+
+        std::ptr::copy_nonoverlapping(ptr, new_ptr, std::cmp::min(layout.size(), new_size));
+        self.dealloc(ptr, layout);
+
+        new_ptr
+    }
```
Comment (tobz) [inline src/internal_telemetry/allocations/allocator/tracing_allocator.rs:115]: Doesn't seem like we need to be overriding this? Especially given that it's omitting the null check to make sure the new allocation succeeded, which we definitely want to keep around.

**Review by tobz — CHANGES_REQUESTED** [review]
I think I _sort_ of see what you're doing, but admittedly, I think it needs some cleanup first to make things clearer before I can pass judgment.

File: src/internal_telemetry/allocations/allocator/tracing_allocator.rs (hunk 4 above)
Comment (arshiyasolei) [reply src/internal_telemetry/allocations/allocator/tracing_allocator.rs:115]: Overriding is mainly for the benefit of inlining. We also need to decide if we want to handle nulls. Currently we panic in case of receiving a null ptr for wrapped allocations.

Comment (tobz) [reply src/internal_telemetry/allocations/allocator/tracing_allocator.rs:115]: But if we aren't tracking allocations, we never call into `get_wrapped_allocation`, we just call `self.allocator.alloc(...)` directly.

I think I'd rather lose the slight performance boost to keep safety invariants in place.

Comment (arshiyasolei) [reply src/internal_telemetry/allocations/allocator/tracing_allocator.rs]: I ended up removing the aborting part

Comment (arshiyasolei) [reply src/internal_telemetry/allocations/allocator/tracing_allocator.rs]: now we return null pointer if allocation failed

File: src/internal_telemetry/allocations/allocator/tracing_allocator.rs [hunk 5]
```diff
@@ -26,62 +26,56 @@ impl<A, T> GroupedTraceableAllocator<A, T> {
     }
 }
 
-impl<A: GlobalAlloc, T: Tracer> GroupedTraceableAllocator<A, T> {
-    #[inline(always)]
-    unsafe fn get_wrapped_allocation(&self, object_layout: Layout) -> (*mut u8, *mut u8, Layout) {
+unsafe impl<A: GlobalAlloc, T: Tracer> GlobalAlloc for GroupedTraceableAllocator<A, T> {
+    #[inline]
+    unsafe fn alloc(&self, object_layout: Layout) -> *mut u8 {
+        if !TRACK_ALLOCATIONS.load(Ordering::Relaxed) {
```
Comment (tobz) [inline src/internal_telemetry/allocations/allocator/tracing_allocator.rs:32]: I think I'd want this to be `Acquire`, along with the callsite where we check if allocation tracking is enabled.

This would just provide a stronger guarantee that things are ordered as we expect / need them to be for our invariants to hold.

Alternatively, maybe this is fine as `Relaxed` because when Vector starts, only the main thread exists, and so there shouldn't yet be any computation that would lead to interleaved load/stores where the ordering would matter... but I'd have to think about that one a little more because I'm not as much of an expert in the memory ordering to feel confident making that determination in this comment alone. 😅

File: src/internal_telemetry/allocations/allocator/tracing_allocator.rs [hunk 6]
```diff
@@ -26,62 +26,56 @@ impl<A, T> GroupedTraceableAllocator<A, T> {
...
+                group_id_ptr.write(group_id.as_raw());
+                self.tracer.trace_allocation(object_size, group_id);
+            },
+        );
+        actual_ptr
     }
 
     #[inline]
     unsafe fn dealloc(&self, object_ptr: *mut u8, object_layout: Layout) {
+        if !TRACK_ALLOCATIONS.load(Ordering::Relaxed) {
```
Comment (tobz) [inline src/internal_telemetry/allocations/allocator/tracing_allocator.rs:65]: Same comment as above about making this `Acquire` instead of `Relaxed`.

File: src/main.rs [hunk 7]
```diff
@@ -5,9 +5,36 @@ use vector::app::Application;
...
+            .unwrap_or_else(|code| {
+                std::process::exit(code);
+            });
+        let allocation_tracing = opts.root.allocation_tracing;
+        REPORTING_INTERVAL_MS.store(
+            opts.root.allocation_tracing_reporting_interval_ms,
+            Ordering::Relaxed,
+        );
+        drop(opts);
+        use std::sync::atomic::Ordering;
```
Comment (tobz) [inline src/main.rs:28]: Can you just move this `use` statement up with the other one on line 10?

File: src/main.rs [hunk 8]
```diff
@@ -5,9 +5,36 @@ use vector::app::Application;
...
+        STARTUP.store(false, Ordering::Relaxed);
+        if allocation_tracing {
+            // Configure our tracking allocator.
+            init_allocation_tracing();
+        } else {
+            // We assume nothing is allocated on the heap at this point.
+            TRACK_ALLOCATIONS.store(false, Ordering::Relaxed);
```
Comment (tobz) [inline src/main.rs:35]: I think we should make these `Release`, fitting with the other comments to make the loads `Acquire`, unless we can prove that the LLVM memory model, and the circumstances of where these loads/stores are happening (in `main`, no other threads, etc), provide a guarantee that no previous value can be read (i.e. reading `true` during load in the tracking allocator after actually having set it to `false`).

**Comment by arshiyasolei** [comment]
Yep given that we added this new lifetime constraint, inverting the logic should work.

File: src/internal_telemetry/allocations/allocator/tracing_allocator.rs (hunk 5 above)
Comment (arshiyasolei) [reply src/internal_telemetry/allocations/allocator/tracing_allocator.rs:32]: Hmm yeah. I am curious about the potential overhead.

**Comment by tobz** [comment]
This PR looks reasonable to me, although what I'm confused about is how it's showing an improvement for so many tests. 🤔

**Comment by blt** [comment]
> This PR looks reasonable to me, although what I'm confused about is how it's showing an improvement for so many tests. thinking

That... is surprising to me too.

**Review by tobz — APPROVED** [review]
Talking through it with Arshia, I believe this is ready to be merged now.

The overhead when allocation tracing is merged in by default, but not enabled at runtime, is essentially zero, or small enough that the regression detector can't seem to reliably distinguish it from the noise.
